src/mqttcm_connect.c

```diff
@@ -1573,11 +1607,95 @@ int isSubscribeNeeded(char *compname)
 	return 2;
 }
 
+char* GetTopicFromSubcribeId(int subscribeId)
+{
+	comp_topic_name_t* temp = g_head;
+	while (temp != NULL)
+	{
+		if(temp->subscribeId == subscribeId)
+		{
+			return (temp->compName);
+		}
+		temp = temp->next;
+	}
+	return NULL;
+}
+
+void AddSubscriptionIdToList(char *comp, int subscribeId)
+{
+	comp_topic_name_t* temp = g_head;
+	while (temp != NULL)
+	{
+		if(strcmp(temp->compName, comp) == 0)
+		{
+			temp->subscribeId = subscribeId;
+			temp->subscribeOnFlag = 1;
+			MqttCMInfo("The component %s is subscribed to topic %s with subscribeId %d\n", temp->compName, temp->topic, subscribeId);
+			return;
+		}
+		temp = temp->next;
+	}
+}
+
+void insert(char * compName, char* topic)
+{
+
+	if( (compName != NULL) && (topic != NULL) )
+	{
+		comp_topic_name_t* newNode = (comp_topic_name_t*)malloc(sizeof(comp_topic_name_t));
+		if(newNode)
+		{
+			memset(newNode, 0, sizeof(comp_topic_name_t) );
+			strncpy(newNode->compName, compName, sizeof(newNode->compName) - 1);
+			strncpy(newNode->topic, topic, sizeof(newNode->topic) - 1);
+			//Subscribe flag is set to 0 as subscribe is not done yet 
+			newNode->subscribeOnFlag = 0;
+			newNode->next = NULL;
+
+			if(g_head == NULL)
+			{
+				g_head = newNode;
+			}
+			else
+			{
+				comp_topic_name_t* current = g_head;
+				while (current->next != NULL)
+				{
+					current = current->next;
+				}
+				current->next = newNode;
+			}
+		}
+		else
+		{
+			MqttCMError("Memory allocation failed\n");
+		}
+	}
+	else
+	{
+		MqttCMError("The compName or topic name is NULL\n");
+	}
+	return;
+
+}
+
+void printList()
+{
+	MqttCMInfo("Inside print function\n");
+	comp_topic_name_t* current = g_head;
+	while (current != NULL)
+	{
+		MqttCMInfo("compname is %s and topic is %s\n", current->compName, current->topic);
+		current = current->next;
+	}
+}
+
 int mqtt_subscribe(char *comp, char *topic)
 {
 	int rc;
 	if(topic != NULL && comp !=NULL)
 	{
+		//To Avoid resubscribe of the same component again
```

Comment:
Check this log in line 1701 - https://github.com/xmidt-org/mqttConnManager/pull/13/files#diff-7be194d13002359bf2c8fb80620b650ecf123581213f712491c4f42db78bc15bR1701

when isSubscribeNeeded returns 0, it is not subscribed .

Reply:
I think isSubscribeNeeded() check in 1699 is not needed as it is already checking inside AddToSubscriptionList func, please confirm if any other use case here ?

Reply:
This check is to avoid resubscribe for this scenario, if the component is restarted it will sent subscribe method set but mqtt is already subscribed to the topic for the same component . Since the rbus onmessage callbacks are auto registered , when message comes to that topic it will be received in onmessage callback.